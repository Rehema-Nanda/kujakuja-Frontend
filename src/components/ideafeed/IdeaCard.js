import React, { Component } from "react";
import moment from "moment";
import { Row, Button } from "reactstrap";
import PropTypes from "prop-types";
import i18n from "i18next";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import convert from "htmr";

import AppConfig from "../../AppConfig";
import lngOptions from "../../i18n/lngOptions";

const googleTranslateKey = AppConfig.GOOGLE_TRANSLATE_API_KEY;
const googleTranslate = require("google-translate")(googleTranslateKey);

const ideaFeedStarredIcon = require("../../img/ideafeed_starred.svg");
const googleTranslateBadge = require("../../img/google_translate_greyscale.svg");

class IdeaCard extends Component {
    constructor(props) {
        super(props);
        this.state = {
            isDetailsAccordionOpen: false,
            translatedIdea: "",
            originalIdeaLang: "",
        };
    }

    componentDidMount() {
        const { idea } = this.props;
        if (i18n.language !== idea.idea_language) {
            this.translateIdea();
            this.getIdeaOriginalLanguage();
        }
    }

    toggleDetailsAccordion = () => {
        const { isDetailsAccordionOpen } = this.state;
        this.setState({ isDetailsAccordionOpen: !isDetailsAccordionOpen });
    };

    generateTranslationMarkUp = () => {
        const { idea } = this.props;
        return (
            `${i18n.language}-x-mtfrom-${idea.idea_language}`
        );
    };

    getIdeaOriginalLanguage = () => {
        const { idea: { idea_language } } = this.props; // eslint-disable-line camelcase
        const language = lngOptions.find((lng) => lng.value === idea_language); // eslint-disable-line camelcase
        const languageName = language ? language.name : idea_language;// eslint-disable-line camelcase
        this.setState({ originalIdeaLang: languageName });
    }

    translateIdea = () => {
        // eslint-disable-next-line camelcase
        const { idea: { idea, idea_language } } = this.props;
        const currentLang = i18n.language;
        googleTranslate.translate(idea, idea_language, currentLang, (err, translate) => {
            if (err) {
                return;
            }
            this.setState({ translatedIdea: translate.translatedText });
        });
    }

    render() {
        const { idea, searchTag } = this.props;
        const { isDetailsAccordionOpen, translatedIdea, originalIdeaLang } = this.state;

        return (
            <div className="ideafeed-item">
                <div className="ideafeed-item-idea">
                    <Row>
                        <p className="small">
                            {moment.utc(idea.created_at).format("LL")}
                            {" | "}
                            {idea.name}
                            {idea.tags && " |"}
                            {idea.tags && idea.tags.split(",").map((tag) => {
                                return (
                                    <Button
                                        key={tag}
                                        type="button"
                                        className="tag"
                                        value={`#${tag}`}
                                        onClick={searchTag}
                                    >
                                        {`#${tag}`}
                                    </Button>
                                );
                            })}
                        </p>
                        {idea.is_starred && (
                            <div className="ideafeed-item-starred">
                                <img alt="" src={ideaFeedStarredIcon} className="icon" />
                            </div>
                        )}
                    </Row>
                    <div className="idea" lang={this.generateTranslationMarkUp()}>
                        &quot;
                        {convert(translatedIdea || idea.idea)}
                        &quot;
                    </div>
                    {i18n.language !== idea.idea_language
                    && (
                        <div className="translation-container">
                            <div className="text">
                                <span
                                    className="show-original"
                                    role="presentation"
                                    onClick={() => {
                                        this.toggleDetailsAccordion();
                                    }}
                                >
                                    {isDetailsAccordionOpen ? i18n.t("ideafeed.hideOriginal") : i18n.t("ideafeed.showOriginal")}
                                    {" "}
                                    {originalIdeaLang}
                                    {" "}
                                    {i18n.t("ideafeed.idea")}
                                </span>
                            </div>
                            <div className="google-badge">
                                <a href="http://translate.google.com/" rel="noopener noreferrer" target="_blank">
                                    <img src={googleTranslateBadge} alt="" />
                                </a>
                            </div>
                        </div>
                    )}
                </div>

                {i18n.language !== idea.idea_language
                && (
                    <div className="ideafeed-item-details">
                        {isDetailsAccordionOpen
                        && (
                            <div className="ideafeed-item-details-idea">
                                <div className="header">
                                    <span className="header-text">{i18n.t("ideafeed.ideaAsCaptured")}</span>
                                </div>
                                <p className="idea show-original-idea">
                                    &quot;
                                    {convert(idea.idea)}
                                    &quot;
                                </p>
                            </div>
                        )}
                        <div
                            className="toggle"
                            role="presentation"
                            onClick={() => {
                                return this.toggleDetailsAccordion();
                            }}
                        >
                            {isDetailsAccordionOpen
                                ? (
                                    <span>
                                        {i18n.t("ideafeed.close")}
                                        {" "}
                                        <FontAwesomeIcon icon="caret-up" />
                                    </span>
                                ) : (
                                    <span>
                                        {i18n.t("ideafeed.details")}
                                        {" "}
                                        <FontAwesomeIcon icon="caret-down" />
                                    </span>
                                )}
                        </div>
                    </div>
                )}
            </div>
        );
    }
}
IdeaCard.propTypes = {
    idea: PropTypes.objectOf(
        PropTypes.oneOfType([
            PropTypes.string,
            PropTypes.bool]),
    ).isRequired,
    searchTag: PropTypes.func.isRequired,
};
export default IdeaCard;
