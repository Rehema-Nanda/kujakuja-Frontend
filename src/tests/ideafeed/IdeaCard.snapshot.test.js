import React from "react";
import { mount } from "enzyme";
import IdeaCard from "../../components/ideafeed/IdeaCard";

describe("IdeaCard component", () => {
    const props = {
        idea: {
            idea: "some random idea",
            created_at: "2019-11-18T14:35:38.000Z",
            is_starred: false,
            name: "Nyabiheke camp",
            tags: "TestTag1,TestTag2",
            idea_language: "en",
        },
        searchTag: jest.fn(),
    };
    const wrapper = mount(<IdeaCard {...props} />);

    it("renders correctly", () => {
        expect(wrapper.debug()).toMatchSnapshot();
    });

    it("should call search function", () => {
        wrapper.find(".tag").first().simulate("click");
        expect(props.searchTag).toHaveBeenCalled();
    });
});
