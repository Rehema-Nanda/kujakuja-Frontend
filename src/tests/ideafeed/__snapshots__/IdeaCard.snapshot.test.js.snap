// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdeaCard component renders correctly 1`] = `
"<IdeaCard idea={{...}} searchTag={[Function: mockConstructor]}>
  <div className=\\"ideafeed-item\\">
    <div className=\\"ideafeed-item-idea\\">
      <Row tag=\\"div\\">
        <div className=\\"row\\">
          <p className=\\"small\\">
            November 18, 2019
             | 
            Nyabiheke camp
             |
            <Button type=\\"button\\" className=\\"tag\\" value=\\"#TestTag1\\" onClick={[Function: mockConstructor]} color=\\"secondary\\" tag=\\"button\\">
              <button type=\\"button\\" value=\\"#TestTag1\\" onClick={[Function: bound onClick]} className=\\"tag btn btn-secondary\\" aria-label={{...}}>
                #TestTag1
              </button>
            </Button>
            <Button type=\\"button\\" className=\\"tag\\" value=\\"#TestTag2\\" onClick={[Function: mockConstructor]} color=\\"secondary\\" tag=\\"button\\">
              <button type=\\"button\\" value=\\"#TestTag2\\" onClick={[Function: bound onClick]} className=\\"tag btn btn-secondary\\" aria-label={{...}}>
                #TestTag2
              </button>
            </Button>
          </p>
        </div>
      </Row>
      <div className=\\"idea\\" lang=\\"undefined-x-mtfrom-en\\">
        &quot;
        some random idea
        &quot;
      </div>
      <div className=\\"translation-container\\">
        <div className=\\"text\\">
          <span className=\\"show-original\\" role=\\"presentation\\" onClick={[Function: onClick]}>
             
            English
             
          </span>
        </div>
        <div className=\\"google-badge\\">
          <a href=\\"http://translate.google.com/\\" rel=\\"noopener noreferrer\\" target=\\"_blank\\">
            <img src={{...}} alt=\\"\\" />
          </a>
        </div>
      </div>
    </div>
    <div className=\\"ideafeed-item-details\\">
      <div className=\\"toggle\\" role=\\"presentation\\" onClick={[Function: onClick]}>
        <span>
           
          <FontAwesomeIcon icon=\\"caret-down\\">
            <i className=\\"fa\\" />
          </FontAwesomeIcon>
        </span>
      </div>
    </div>
  </div>
</IdeaCard>"
`;
